ENH: Add array API inspection functions

asmeurer opened the pull request:
These are new functions defined by the array API 2023.12 standard to inspect
details about the array API namespace. See
https://data-apis.org/array-api/latest/API_specification/inspection.html for
more details.

Even though these functions are pretty simple for NumPy, I have a few questions about the implementation:

- Is `numpy/lib/_info.py` the correct to put this?
- For the functions that return dtype objects, is it better to return dtype classes or dtype instances? I used classes for now since that is the simplest. 
- I have assumed here that all this information is always the same for any build of NumPy. But if that isn't the case, let me know so I can change the functions to look up the data dynamically. Specifically are the default dtypes always the same, and does NumPy always define all the array API dtypes?
- I added some tests, but they are mostly trivial, so if anyone has any suggestions on how to improve them let me know.
- I don't know how to add the docs here to Sphinx, other than `__array_namespace_info__`. The issue is that these functions are accessed like `np.__array_namespace_info__().capabilities()`, so it's not a normal dotted module name. I could add the private module to Sphinx. Is it possible to do this without referring to the private API? Another alternative I thought of is to make `__array_namespace_info__` a class so that the functions are just methods on the class (I would have rather preferred if the array API had defined `__array_namespace_info__` as a simple submodule rather than a function, but it may be too late to change that).
- Related to the docs, is there a NumPy docstring style guide or anything like that? I tried to follow the style of other docs as closely as possible. 


<!--         ----------------------------------------------------------------
                MAKE SURE YOUR PR GETS THE ATTENTION IT DESERVES!
                ----------------------------------------------------------------

*  FORMAT IT RIGHT:
      https://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message

*  IF IT'S A NEW FEATURE OR API CHANGE, TEST THE WATERS:
      https://www.numpy.org/devdocs/dev/development_workflow.html#get-the-mailing-list-s-opinion

*  HIT ALL THE GUIDELINES:
      https://numpy.org/devdocs/dev/index.html#guidelines

*  WHAT TO DO IF WE HAVEN'T GOTTEN BACK TO YOU:
      https://www.numpy.org/devdocs/dev/development_workflow.html#getting-your-pr-reviewed
-->

asmeurer committed:
Add array API inspection functions

These are new functions defined by the array API 2023.12 standard to inspect
details about the array API namespace. See
https://data-apis.org/array-api/latest/API_specification/inspection.html for
more details.

asmeurer committed:
Add "max rank" to the capabilities() dictionary

This is new in the unreleased 2024.12 standard, but I don't see an issue with
including it now.

asmeurer committed:
Add documentation to the inspection API functions

Unfortunately, I don't see an easy way to include these docstrings in Sphinx,
other than __array_namespace_info__, because they are hidden behind a private
namespace that is only accessible from np.__array_namespace_info__().

asmeurer committed:
Add basic tests for the array API inspection functions

asmeurer committed:
Fix capabilities()['max rank'] to be 64

asmeurer committed:
Add some consistency checks to the info namespace tests

asmeurer committed:
Remove the unreleased "max rank" from the capabilities() dictionary

It's better to not include unreleased standard things in case they change.

asmeurer committed:
Reformat info files with black

mtsokol commented on numpy/lib/_info.py:
```diff
@@ -0,0 +1,384 @@
+
+from typing import TYPE_CHECKING
+
+from numpy import (
```
Let's import dtypes from `from numpy._core import ...`, to avoid cyclic dependency. I would also move contents of this file to `numpy/lib/introspect.py`.

mtsokol commented on numpy/lib/_info.py:
```diff
@@ -0,0 +1,384 @@
+        )
+    if kind is None:
+        return {
+            "bool": bool,
+            "int8": int8,
+            "int16": int16,
+            "int32": int32,
+            "int64": int64,
+            "uint8": uint8,
+            "uint16": uint16,
+            "uint32": uint32,
+            "uint64": uint64,
+            "float32": float32,
+            "float64": float64,
+            "complex64": complex64,
+            "complex128": complex128,
```
Nitpick: maybe instead of repeating contents of these dicts, let's keep `signed integer`, `unsigned integer`, `bool`, `real`, `complex` dicts, and define the rest of them by merging existing ones: `integral = signed_integer | unsigned_integer` etc.

mtsokol reviewed (changes requested):
My comments:
- I would prefer `__array_namespace_info__` to be a class with methods, as the standard allows it: 
  > In the namespace (or class) returned by __array_namespace_info__,
- We already have `numpy/lib/introspect.py`, maybe we can move the implementation there?
- In terms of testing: let's bump `array-api-tests` CI job to the latest `array-api-tests` version:
  - Bump the hash in: https://github.com/numpy/numpy/blob/f2c93e2aa5c97dd6837efff8bde59304cf5d0851/.github/workflows/linux.yml#L231 to the latest commit that includes `2023.12` standard. I think it's `33f2d2ea2f3dd2b3ceeeb4519d55e08096184149`? (does it test `2023.12` items?)
  - Let's bump the version to `2023.12` in https://github.com/numpy/numpy/blob/f2c93e2aa5c97dd6837efff8bde59304cf5d0851/numpy/__init__.py#L292
  - I think there might be needed some slight adjustments to the `array-api-skips.txt` file.
- I think returning dtype instances are fine.
- I think the default dtypes should be always the same. `real` and `complex` are straightforward. `integral` and `indexing` are a bit tricky. Here's some discussion about it https://github.com/numpy/numpy/pull/24224. I think for `integral` and `indexing` it should be `np.intp`.

rgommers commented:
> We already have `numpy/lib/introspect.py`, maybe we can move the implementation there?

That doesn't sound quite right to me, both because that's a public namespace and because this inspection functionality will not be part of `numpy.lib` but live in the main namespace. I suggest `numpy/_array_api_info.py`.

This should have dedicated tests, not only rely on `array-api-tests`. It can be simple. Two options for a test for are `numpy/tests/` or `numpy/_core/tests/`. I think I prefer the latter, since that's where most tests for the main namespace live.



> I think returning dtype instances are fine.

Agreed, that's more idiomatic.

> I think for `integral` and `indexing` it should be `np.intp`.

That sounds right to me.

> Related to the docs, is there a NumPy docstring style guide or anything like that?

https://numpydoc.readthedocs.io/en/latest/format.html

mtsokol commented:
> That doesn't sound quite right to me, both because that's a public namespace and because this inspection functionality will not be part of `numpy.lib` but live in the main namespace. I suggest `numpy/_array_api_info.py`.

Right, good point.

> This should have dedicated tests, not only rely on `array-api-tests`. It can be simple. Two options for a test for are `numpy/tests/` or `numpy/_core/tests/`. I think I prefer the latter, since that's where most tests for the main namespace live.

Sure - I think these new functions should have a separate test suite, but `array-api-tests` CI can also be updated here.

asmeurer commented:
array-api-tests doesn't currently support the inspection APIs. I've added support for testing the signatures at https://github.com/data-apis/array-api-tests/pull/262, but I still need to implement tests for the actual functionality.

asmeurer commented:
> Let's bump the version to 2023.12 in

Better not do that until we have full 2023 support. There are still a few other things that need to be updated.

asmeurer committed:
Change __array_namespace_info__ to a class

This makes the functions on the namespace easier to access for things like
Sphinx.

asmeurer committed:
Rename numpy/lib/_info.py to numpy/_array_api_info.py

asmeurer committed:
Change the default integral and indexing dtypes from int64 to intp

asmeurer committed:
Update default_dtypes() test to be more indirect

asmeurer committed:
Change the array API inspection functions to return dtype instances

asmeurer committed:
Add inspection functions to Sphinx

asmeurer commented:
I've added the functions to Sphinx. Unfortunately, the pydata-sphinx-theme does not properly word wrap function names in several places:

<img width="1074" alt="image" src="https://github.com/numpy/numpy/assets/71486/f5fa41ef-25d4-462e-a7ad-e198aed25017">
<img width="829" alt="image" src="https://github.com/numpy/numpy/assets/71486/602b24af-1453-4932-80fb-575c513471a3">

(see also https://github.com/scipy/scipy/issues/20635). This will need to be fixed in the upstream theme. 

I am pretty sure some of my doctests are wrong, but I have no idea how to actually run them. The refcheck_guide script doesn't seem to run the doctests for these functions.

asmeurer committed:
Reformat files with black

asmeurer committed:
Set __module__ = 'numpy' on __array_namespace_info__

asmeurer committed:
Fix linter errors

asmeurer commented:
I'm not sure about the typing stuff here. I mainly copied it from the spec, changing some generic types like `Device` to be more specific like `str`.

mtsokol reviewed (commented):
Only one comment. I can check docs errors, could you rebase your branch to the latest `main`?

asmeurer committed:
Fix test for default indexing and integer dtypes

asmeurer committed:
Fix docs build errors

asmeurer committed:
Merge branch 'main' into array-api-inspection

asmeurer commented:
Issues should be fixed.

mtsokol committed:
Create typing stub

mtsokol committed:
Add missing default values

mtsokol reviewed (approved):
@asmeurer I added typing stubs file and typing tests. Now docs build without warnings. LGTM, but as I pushed some changes we rather need one more person to take a look.

mattip commented:
Thanks @asmeurer , @mtsokol. We can iterate on this as needed.
